fix(chips): improve chips accessibility

tinayuangao opened the pull request:
- Changes chip role to gridcell
 - Changes chip list role to grid
 - Allow keyboard navigation from/to chip input

crisbeto commented on src/lib/chips/chip-input.ts:
```diff
@@ -21,6 +33,9 @@ export interface MatChipInputEvent {
   value: string;
 }
 
+export class MatChipInputBase {}
```
Can you mark these as `@docs-private`? AFAIK this one may end up in the docs.

crisbeto commented on src/lib/chips/chip-input.ts:
```diff
@@ -124,8 +146,18 @@ export class MatChipInput {
+      const isRtl = this._dir && this._dir.value === 'rtl';
+      const isPreviousKey = event &&
+        (event.keyCode === LEFT_ARROW && !isRtl) ||
+        (event.keyCode === RIGHT_ARROW && isRtl);
```
Doesn't the key manager handle the previous key already?

tinayuangao replied:
No. The active item index is -1.

tinayuangao commented:
Current chip list supports up/down arrows between chips. I didn't add them in chip input because chip input may be associated with an autocomplete dropdown. Up/down arrows may be used to move between items inside the autocomplete dropdown, which is confusing if that key event also moves between chip input and chips.

crisbeto commented:
I meant more that the user might expect that pressing up/down arrow would move them down a "row". E.g. currently the datepicker's calendar uses `grid` and NVDA reads out "Row X, Column Y".

tinayuangao commented:
@crisbeto The chip list will have only one row, containing all chips.

@jelbourn 
- Updated `chips.md`
- `role="row"` is added to `chip-list` template. For standalone chips, added instruction in `chips.md`
- `aria-label` should be used to let user know the chip is removable. Added to chip component and `chips.md`
- `aria-readonly` set to true for `chip`, since chips are not editable
- Not sure the value `aria-readonly` for chip list (grid). Since the chips may be removable or there may be a chip input.

mmalerba commented on src/lib/chips/chip-input.ts:
```diff
@@ -124,8 +148,18 @@ export class MatChipInput {
     if (!this._inputElement.value && !!event) {
-      this._chipList._keydown(event);
+      const isRtl = this._dir && this._dir.value === 'rtl';
+      const isPreviousKey = event &&
```
I think the parens are wrong on this expression. I would change to:

`event && event.keyCode === (isRtl ? RIGHT_ARROW : LEFT_ARROW);`

mmalerba commented on src/lib/chips/chip-list.ts:
```diff
@@ -467,11 +466,17 @@ export class MatChipList extends _MatChipListMixinBase implements MatFormFieldCo
   _keydown(event: KeyboardEvent) {
     const target = event.target as HTMLElement;
+    const isRtl = this._dir && this._dir.value === 'rtl';
+    const isNextKey = (event.keyCode === RIGHT_ARROW && !isRtl) ||
```
This is more concise: `event.keyCode === (isRtl ? LEFT_ARROW : RIGHT_ARROW)`

tinayuangao commented:
@mmalerba Thanks for review. Comments addressed.

jelbourn commented:
Were you able to get the delete key working under VoiceOver?

tinayuangao commented:
@jelbourn I can use the delete key to remove the chips, but the VoiceOver will not know there's a delete icon unless user set the aria-label to describe it.

mmalerba commented:
@tinayuangao shouldn't 

```html
<mat-icon matChipRemove>
```

really be 

```html
<button mat-icon-button matChipRemove><mat-icon>remove</mat-icon></button>
```

In this case it's more obvious that `aria-label` is needed since it's a `<button>`

tinayuangao commented:
@jelbourn @mmalerba 
Updated the styles so developers can put `button mat-icon-button matChipRemove><mat-icon>remove</mat-icon></button>`.

tinayuangao commented:
@jelbourn Changed docs and added `aria-label` to chip-list. Please take a look. Thanks!
